Add first pass at GuestArray

kubkon opened the pull request:
While this is a WIP, I thought it might be useful to initiate a draft PR so that you can monitor the progress and shout out if you find anything amiss, wrong, etc.

kubkon commented on crates/memory/src/memory.rs:
```diff
@@ -359,3 +359,72 @@ impl<'a, T> Drop for GuestRefMut<'a, T> {
+        // Then, extract as slice
+        Ok(unsafe {
+            std::slice::from_raw_parts(self.ptr.as_raw() as *const T, self.num_elems as usize)
+        })
```
I think this is slightly better now compared to what it was in the previous commit d910c0e. I hope I'm right in my thinking here that by firstly validating that all (expected) pointers are valid, we can then safely extract the slice `&'a [T]` utilising `std::slice::from_raw_parts` with a raw pointer cast.

pchickey replied:
I believe this is correct as well.

kubkon commented on crates/memory/src/memory.rs:
```diff
@@ -359,3 +359,267 @@ impl<'a, T> Drop for GuestRefMut<'a, T> {
+where
+    T: GuestType,
+{
+    r#ref: GuestRef<'a, T>,
```
SGTM! I'll make sure to follow this style then from now on.

kubkon replied:
Done in 360aa71.

kubkon commented on crates/memory/src/memory.rs:
```diff
@@ -359,3 +359,267 @@ impl<'a, T> Drop for GuestRefMut<'a, T> {
+{
+    ptr: GuestPtr<'a, T>,
+    num_elems: u32,
+}
```
Excellent point, thanks! I think I'm also partial towards the latter. Let me see what I can conjure up!

kubkon replied:
Done in cc4f33f.

kubkon commented on crates/memory/src/memory.rs:
```diff
@@ -359,3 +359,267 @@ impl<'a, T> Drop for GuestRefMut<'a, T> {
+{
+    ptr: GuestPtrMut<'a, T>,
+    num_elems: u32,
+}
```
Done in cc4f33f.

kubkon commented:
While addressing the comments (thanks!), I've gone ahead and implemented `fmt::Debug` for memory types. Note thought that the implementation is *not* enforced on the generic parameter `T` in the struct; rather, if `T` impls `fmt::Debug` then so does the memory type such as `GuestPtr<'_, T>`. This should simplify testing and debugging in general (I hope!).
